Two solutions for pathSum and driver (#187)
UWB-ACM/CTCI: 2020-01-Winter/4_trees/README.md

<a name="p1"/>

### 1. Path Sum

Source: [Leetcode] https://leetcode.com/problems/path-sum/

#### Scenario

Problem Statement: Given a binary tree and a sum, determine if the tree has a root-to-leaf path such that adding up all the values along the path equals the given sum.

Note: A leaf is a node with no children.

#### Example Input
Example 1:

Given the below binary tree and sum = 22,

>       5

Reviewer: ````suggestion
```
      5
````

     / \
    4   8
   /   / \
  11  13  4
 /  \      \

> 7    2      1

Reviewer: ````suggestion
7    2      1
```
````

> return true, as there exist a root-to-leaf path 5->4->11->2 which sum is 22.

Reviewer: For any sections where monospaced formatting matters, I _strongly_ recommend enclosing the lines in backticks.

Right now, the file looks like:
![funky_formatting](https://user-images.githubusercontent.com/29528527/74423123-97182880-4e04-11ea-9399-8071254a088f.png)

By enclosing the tree segment in 3 backticks, you'll get the structure of the text preserved, including leading spaces. I will add them as suggestions you can commit.

Example 2:
Given the below binary tree and sum = 1,

>       1

Reviewer: Moar backticks
````suggestion
```
      1
````

     / \

>     -2  3

Reviewer: ````suggestion
   -2   3
```
````

return false, as there does not exist a root-to-leaf path which sums 1.

#### Function Signature
Example C++ function signature:

> /**

Reviewer: Without enclosing code blocks in triple backticks, you will wind up with some very strange rendering on GitHub's part. This happens because the code-specific characters, indentation, etc are getting recognized as Markdown formatting instead of a code block. Here's what it looks like before the changes:

![code_block](https://user-images.githubusercontent.com/29528527/74423510-448b3c00-4e05-11ea-81d1-f1a070fed561.png)

````suggestion
```c++
/**
````

 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Solution {
public:
    bool hasPathSum(TreeNode* root, int sum) {

    //your code here
}

> };

Reviewer: Close the code block
````suggestion
};
```
````

<!-- Don't remove -->
<a name="s1"/>

### 1. Path Sum

> Source: [Leetcode]https://leetcode.com/problems/path-sum/

Reviewer: Nitpick: markdown hyperlink formatting
```suggestion
Source: [Leetcode](https://leetcode.com/problems/path-sum/)
```

#### Naive/Simple Solution
This implementation uses the queue for BFS iterating over the nodes.
**Time Complexity**: O(n) where N is the number of nodes in the tree.

> class Solution {

Reviewer: Without triple backticks for code block formatting, the code gets rendered as follows:
![code](https://user-images.githubusercontent.com/29528527/74424026-1528ff00-4e06-11ea-893e-5d6d924b052b.png)

Include `c++` as the language after the opening backtick section so you get syntax highlighting on top of code block formatting.
````suggestion
```c++
class Solution {
````

public:
 bool hasPathSum(TreeNode* root, int sum) {
  //gone through the tree, no path
  if (root == NULL)
	  return false;
 
  queue <TreeNode*> qlayer;
  qlayer.push(root);
  while (!qlayer.empty())
  {
	  TreeNode *ptr= qlayer.front();
	  //if no children
	  if (ptr->left == NULL && ptr->right == NULL)
	  {
		  //value at the end is equal to sum
		  //means there is a path from root to leaf 
		  if (ptr->val==sum)
			  return true;
	  }
	  else
	  {
		  //go layer by layer
		  //the current value is added to the left child
		  //add the node to the queue
		  if (ptr->left != NULL)
		  {
			  ptr->left->val += ptr->val;
			  qlayer.push(ptr->left);
		  }
		  //the current value is added to the right child
		  if (ptr->right != NULL)
		  {
			  ptr->right->val += ptr->val;
			  qlayer.push(ptr->right);
		  }
	  }
	  qlayer.pop();
  }
  return false;
 }

> };

Reviewer: Close the code block
````suggestion
};
```
````

#### Optimal Solution
This implementation uses recursion  for DFS.

> **Time Complexity**: O(2^n) the branch factor is two in a binary search tree and N is the number of times the function is called.

Reviewer: Good job thinking about complexity tradeoffs here; double-check what you want `N` to be vs what you want the total complexity to represent. I think the total complexity will represent the number of times the function is called, whereas `N` is the number of nodes that have to be processed and perform recursive calls.

```suggestion
**Time Complexity**: `O(2^n)` -- the branch factor is two in a binary search tree and N is the number of nodes in the tree.
```

> class Solution {

Reviewer: Add backticks to format code in a code block; current formatting gets rendered like the following example:
![code](https://user-images.githubusercontent.com/29528527/74423864-ca0eec00-4e05-11ea-9dc0-026edcb0031d.png)

````suggestion
```c++
class Solution {
````

public:
	bool hasPathSum(TreeNode* root, int sum) {
		//gone through the tree, no path
		if (root == NULL)
			return false;
		//subtract root from the sum at each visited node
		sum -= root->val;
		//if no children
		if (root->left == NULL && root->right == NULL)
		{
			//sum after subtracting is equal to 0
			//means there is a path from root to leaf 
			if (sum == 0)
				return true;
		}
		//recursive call passing left child or right child as root and new sum
		return (hasPathSum(root->left, sum) || hasPathSum(root->right, sum));
	}

> };

Reviewer: Close code block
````suggestion
};
```
````
